### ISSUE TITLE
Snapshot files should be sanitized before being selected by a CCF node at startup

### ISSUE THREAD

**Issue by andpiccione** [microsoft/CCF#6196]
**Describe the bug**

We came across a scenario where a CCF node may start from a "blank" snapshot, [forcing the node to start from ledger files](https://github.com/microsoft/CCF/blob/main/src/host/main.cpp#L677-L700) instead and increasing the node startup time. As per [this method](https://github.com/microsoft/CCF/blob/8ced7954433ff2cedd4549244c3c64737b79ba89/src/host/snapshots.h#L141), CCF is choosing the latest snapshot file in the target snapshot directory based on the file name only (picking the one with the highest sequence number); it would be sensible to have a few sanity checks in place to verify that a candidate snapshot file is indeed valid, or else ignore it and look for the newest valid snapshot in the directory.

**To Reproduce**

1. Start a CCF network
2. Configure a new node to join the main network, making it use a snapshot directory containing a 0-byte snapshot file with the highest sequence number. Upon startup, the CCF node loads the 0-byte snapshot, but it may eventually end up replaying transactions from the ledger files because the snapshot is empty / not usable.

**Expected behavior**

Before selecting a snapshot for a join/recovery node, CCF should check that the file is valid (e.g., either by running a few simple sanity checks or by directly loading and parsing the file). If the file is deemed not valid, the file should be skipped and the validation process should be repeated for the next candidate snapshot in the directory until a suitable one is found.

**Environment information**

Tested on an Azure Managed CCF instance running on SGX platform, using a [CCF 5.0.0-dev15 build](https://github.com/microsoft/CCF/releases/tag/ccf-5.0.0-dev15) from the CCF `mcr.microsoft.com:ccf/app/run-js:5.0.0-dev15-sgx` runtime image.

**Additional context**

We have seen a few occurrences of a 0-byte snapshot file in the shared ledger directory, which presumably originates from a concurrent copy of the file from the ledger volume to the shared volume (using `azcopy`)  while it's still being written to by the node (rare situation, but it could explain why we haven't seen many occurrences so far). We are fixing the error on our side by overwriting the file in the shared volume during the copy if the source file is newer, but we were thinking it could also make sense for CCF to do a quick "sanity check" of a candidate snapshot to select, to ensure it is valid.

**Comment by achamayou** [comment]
@andpiccione I will try to reproduce this, but [find_latest_committed_snapshot_in_directory](https://github.com/microsoft/CCF/blob/8ced7954433ff2cedd4549244c3c64737b79ba89/src/host/snapshots.h#L141) only selects a file path, and does not look inside. So the branch [you point to](https://github.com/microsoft/CCF/blob/main/src/host/main.cpp#L677-L700) to start from the ledger files should not be taken so long as the file exists, even if it is empty.

The snapshot is passed into the enclave, later deserialised and integrity-checked during node creation, quite a few levels further. It's possible to move this earlier, but not that close because it needs to stay trusted. If we do this on the host side, we need to duplicate it.

On SNP, where the whole binary is in the TEE, we can move this arbitrarily, and validate the snapshot the first time we see it, which would be quite convenient.

**Comment by andpiccione** [comment]
> I will try to reproduce this, but [find_latest_committed_snapshot_in_directory](https://github.com/microsoft/CCF/blob/8ced7954433ff2cedd4549244c3c64737b79ba89/src/host/snapshots.h#L141) only selects a file path, and does not look inside. So the branch [you point to](https://github.com/microsoft/CCF/blob/main/src/host/main.cpp#L677-L700) to start from the ledger files should not be taken so long as the file exists, even if it is empty.

@achamayou I'm not familiar with the codebase or C++ so I might as well be mistaken, but [here](https://github.com/microsoft/CCF/blob/6ec049e4b99426de7f503ff96288b562d4a37a53/src/host/main.cpp#L688C9-L688C26) you are reading the contents of the file returned by [find_latest_committed_snapshot](https://github.com/microsoft/CCF/blob/6ec049e4b99426de7f503ff96288b562d4a37a53/src/host/snapshots.h#L304) and assign it to the `startup_snapshot` variable (initialized to an [empty vector before the `if` condition](https://github.com/microsoft/CCF/blob/6ec049e4b99426de7f503ff96288b562d4a37a53/src/host/main.cpp#L677) you referred to). If the file is empty, I believe the variable would be assigned an empty vector [here](https://github.com/microsoft/CCF/blob/6ec049e4b99426de7f503ff96288b562d4a37a53/src/host/main.cpp#L688); this means that In either cases (whether a snapshot is not found or the file is empty) the `startup_snapshot` would contain the same empty vector at the end of the branching. But again, I might miss something obvious here, maybe they do have different values even if the file is empty and they are handled differently downstream somehow.

> The snapshot is passed into the enclave, later deserialised and integrity-checked during node creation, quite a few levels further. It's possible to move this earlier, but not that close because it needs to stay trusted. If we do this on the host side, we need to duplicate it.
On SNP, where the whole binary is in the TEE, we can move this arbitrarily, and validate the snapshot the first time we see it, which would be quite convenient.

I think even some basic sanity checks without deserialization would suffice, like checking if the file is empty. Even though fully parsing the file would be ideal to ensure the format is valid.

**Comment by andpiccione** [comment]
@achamayou Makes sense, I think adding a simple check for the file size (as well as any other sensible check that could be done without reading the file contents) and continue to search for a non-empty snapshot would be a good start. We may be able to improve the sanitization by actually deserializing the snapshot the first time we see it once we fully move to SNP.

**Comment by achamayou** [comment]
#6859 added logic to ignore empty snapshot files on startup.
